docs/security.md

```diff
@@ -0,0 +1,199 @@
+the image is a visual representation of the trust boundary between the
+inherently untrusted execution context of the annotated page, and the trusted
+execution context of the client frame.
+
+Instead, the components of the client which execute in the annotated page must
+communicate with the client frame using [cross-document
+messaging](https://en.wikipedia.org/wiki/Web_Messaging). It is important that
+such **cross-document messaging should expose only the minimum information
+necessary about user data** to code executing in the annotated page. For
+example: in order to draw highlights, the annotated page needs to know the
+location of annotations, but it does not ever need to know the body text of an
+an annotation, and so it should not be possible to expose this over the
+messaging interface.
+
+_TODO 2017-03-08: currently the client shares an origin with the annotation
+service when delivered by any mechanism other than the Chrome extension. This
+makes any XSS vulnerability in the client a problem for the service and vice
+versa. We need to move the client to its own origin to better isolate the client
+from the service and minimise the risk posed by XSS._
+
+### Input sanitisation
+
+As alluded to above, the client frame is a trusted execution context. Any code
+running there has full access to everything the user has access to, which may
+constitute a major security flaw if that code was provided by another user (say,
+as a `<script>` tag in the body of an annotation).
+
+This is an example of a cross-site scripting attack (XSS) and must be mediated
+by ensuring that **any and all user content displayed in the client frame is
+appropriately escaped and/or sanitised**.
+
+### Transport Layer Security
+
+We ensure that it is hard to eavesdrop on traffic between the client and the
+annotation service by communicating with the annotation service over encrypted
+channels (`https://` and `wss://`).
+
+_TODO 2017-03-08: This is not currently enforced by the client. Perhaps
+production builds of the client should refuse to communicate with annotation
+services over insecure channels?_
+
+### Clickjacking protections
+
+The most straightforward way to protect an application from most kinds of
+clickjacking is the [`frame-ancestors` Content-Security-Policy
+directive](https://w3c.github.io/webappsec-csp/#directive-frame-ancestors) or
+the older [`X-Frame-Options` HTTP
+Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options).
+
+Unfortunately, the client runs in a framed context (and on arbitrary origins) by
+default, so simply applying `X-Frame-Options: DENY` would break the client
+entirely.
+
+_TODO 2017-03-08: The Hypothesis client would appear to have very little
+protection against clickjacking attacks that allow arbitrary websites to trick
+Hypothesis users into performing actions they did not intend to perform. It's
+not immediately clear what tools we have at our disposal to solve this problem._
```

Comment:
This document is really helpful to me in understanding why the client works the way it does.  Thanks.

I think it's great to have these out in the open not only because you of course want a secure service, but also because when less experienced coders like me are unsure of what the main trust breaches are, they are most likely to become overly conservative. When you have more certain of how to prevent a breach, you can be a little more innovative.  So thanks for addition!